pgp/keysource.go

```diff
@@ -0,0 +1,202 @@
+	return time.Since(key.CreationDate).Hours() > 24*30*6
+}
+
+func (key *GPGMasterKey) ToString() string {
+	return key.Fingerprint
+}
+
+func (key *GPGMasterKey) gpgHome() string {
+	dir := os.Getenv("GNUPGHOME")
+	if dir == "" {
+		usr, err := user.Current()
+		if err != nil {
+			return "~/.gnupg"
+		}
+		return path.Join(usr.HomeDir, ".gnupg")
+	}
+	return dir
+}
+
+func NewGPGMasterKeyFromFingerprint(fingerprint string) GPGMasterKey {
+	return GPGMasterKey{
+		Fingerprint:  strings.Replace(fingerprint, " ", "", -1),
+		CreationDate: time.Now(),
+	}
+}
+
+func GPGMasterKeysFromFingerprintString(fingerprint string) []GPGMasterKey {
+	var keys []GPGMasterKey
+	if fingerprint == "" {
+		return keys
+	}
+	for _, s := range strings.Split(fingerprint, ",") {
+		keys = append(keys, NewGPGMasterKeyFromFingerprint(s))
+	}
+	return keys
+}
+
+func (key *GPGMasterKey) loadRing(path string) (openpgp.EntityList, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return openpgp.EntityList{}, err
+	}
+	defer f.Close()
+	keyring, err := openpgp.ReadKeyRing(f)
+	if err != nil {
+		return keyring, err
+	}
+	return keyring, nil
+}
+
+func (key *GPGMasterKey) secRing() (openpgp.EntityList, error) {
+	return key.loadRing(key.gpgHome() + "/secring.gpg")
+}
+
+func (key *GPGMasterKey) pubRing() (openpgp.EntityList, error) {
+	return key.loadRing(key.gpgHome() + "/pubring.gpg")
+}
+
+func (key *GPGMasterKey) fingerprintMap(ring openpgp.EntityList) map[string]openpgp.Entity {
+	fps := make(map[string]openpgp.Entity)
+	for _, entity := range ring {
+		fp := strings.ToUpper(hex.EncodeToString(entity.PrimaryKey.Fingerprint[:]))
+		if entity != nil {
+			fps[fp] = *entity
+		}
+	}
+	return fps
+}
+
+func (key *GPGMasterKey) passphrasePrompt(keys []openpgp.Key, symmetric bool) ([]byte, error) {
+	conn, err := gpgagent.NewConn()
+	if err == gpgagent.ErrNoAgent {
+		fmt.Println("gpg-agent not found, continuing with manual passphrase input...")
+		fmt.Print("Enter PGP key passphrase: ")
+		pass, err := gopass.GetPasswd()
+		if err != nil {
+			return nil, err
+		}
+		for _, k := range keys {
+			k.PrivateKey.Decrypt(pass)
+		}
+		return pass, err
+	}
+	if err != nil {
+		return nil, fmt.Errorf("Could not establish connection with gpg-agent: %s", err)
+	}
+	defer conn.Close()
+	for _, k := range keys {
+		req := gpgagent.PassphraseRequest{
+			CacheKey: k.PublicKey.KeyIdShortString(),
+			Prompt:   "Passphrase",
+			Desc:     fmt.Sprintf("Unlock key %s to decrypt sops's key", k.PublicKey.KeyIdShortString()),
+		}
+		pass, err := conn.GetPassphrase(&req)
+		if err != nil {
+			return nil, fmt.Errorf("gpg-agent passphrase request errored: %s", err)
+		}
+		k.PrivateKey.Decrypt([]byte(pass))
+		return []byte(pass), nil
+	}
+	return nil, fmt.Errorf("No key to unlock")
+}
+
+func (key GPGMasterKey) ToMap() map[string]string {
+	out := make(map[string]string)
+	out["fp"] = key.Fingerprint
+	out["created_at"] = key.CreationDate.Format(DateFormat)
+	out["enc"] = key.EncryptedKey
```

Comment:
I'm putting the comment here but it applies to the entire metadata struct: why does it need to be a map? We need to treat the file data as `interface{}` because we don't know its structure in advance, but the metadata branch is deterministic. It would be much nicer to handle the metadata branch as structs instead. The `sops` package would define the base struct and `pgp` and `kms` packages would define the struct for their own keys.